shared/src/sources/balancer/graph_api.rs

```diff
@@ -55,8 +58,8 @@ impl BalancerSubgraphClient {
         while {
             let page = self
                 .0
-                .query::<pools_query::Data>(
-                    pools_query::QUERY,
+                .query::<pools_query::Data<WeightedToken>>(
+                    pools_query::WEIGHTED_POOL_QUERY,
```

Comment:
I'd like to refactor the common part out:
```rust
    // We do paging by last ID instead of using `skip`. This is the
    // suggested approach to paging best performance:
    // <https://thegraph.com/docs/graphql-api#pagination>
    async fn i_dont_have_a_good_name_for_this<T: DeserializeOwned>(
        &self,
        block_number: u64,
        query: &str,
    ) -> Result<Vec<PoolData<T>>> {
        let mut result = Vec::new();
        let mut last_id = H256::default();
        while {
            let page = self
                .0
                .query::<pools_query::Data<T>>(
                    query,
                    Some(json_map! {
                        "block" => block_number,
                        "pageSize" => QUERY_PAGE_SIZE,
                        "lastId" => json!(last_id),
                    }),
                )
                .await?
                .pools;

            let has_next_page = page.len() == QUERY_PAGE_SIZE;
            if let Some(last_pool) = page.last() {
                last_id = last_pool.id;
            }
            result.extend(page);
            has_next_page
        } {}
        Ok(result)
    }

    /// Retrieves the list of registered stable pools from the subgraph.
    pub async fn get_registered_pools(&self) -> Result<RegisteredPools> {
        let block_number = self.get_safe_block().await?;

        let pools_0 = self
            .i_dont_have_a_good_name_for_this::<WeightedToken>(
                block_number,
                pools_query::WEIGHTED_POOL_QUERY,
            )
            .await?
            .into_iter()
            .map(|pool| (pool.factory, pool.into_registered_pool(block_number)));
        let pools_1 = self
            .i_dont_have_a_good_name_for_this::<StableToken>(
                block_number,
                pools_query::STABLE_POOL_QUERY,
            )
            .await?
            .into_iter()
            .map(|pool| (pool.factory, pool.into_registered_pool(block_number)));

        let mut pools_by_factory = HashMap::<H160, Vec<RegisteredPool>>::new();
        for (factory, pool) in pools_0.chain(pools_1) {
            let pool = match pool {
                Ok(pool) => pool,
                Err(err) => todo!(),
            };
            pools_by_factory
                .entry(factory.unwrap_or_default())
                .or_default()
                .push(pool);
        }

        Ok(RegisteredPools {
            fetched_block_number: block_number,
            pools_by_factory,
        })
    }
```

Reply:
Good eye and thanks for the suggestion. I have been trying to think of a way to do this and hadn't yet got back to it.